### PR TITLE
add niceness options to restic

### PR DESCRIPTION
**Pull request by ibizaman**
## Description of changes

Adds `nice` and `ionice` support for restic. Inspired by:
- https://restic.readthedocs.io/en/latest/faq.html#how-to-prioritize-restic-s-io-and-cpu-time
- https://github.com/HubbeKing/restic-kubernetes/blob/master/backup.sh#L6
- https://github.com/NixOS/nixpkgs/blob/8610e082e04a6dde08af882fbaac74723ae75d5e/nixos/modules/services/backup/btrbk.nix#L102-L111

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [x] [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - [ ] and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - [ ] or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - [ ] made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`) **(Tested on my server)**
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/services/backup/restic.nix [hunk 1]
```diff
@@ -256,6 +256,24 @@ in
             having to manually specify most options.
           '';
         };
+
+        niceness = mkOption {
+          description = lib.mdDoc "Niceness for instances of restic. Also applies to remote ones connecting via ssh when positive. Only used for `restic backup`, `restic forget` and `restic check` commands.";
+          type = types.ints.between (-20) 19;
+          default = 10;
```
Comment (oxzi) [inline nixos/modules/services/backup/restic.nix:263]: This changes the current behavior and would give restic way more CPU priority, resulting in a surprising change for some. Thus, I would set the `default` value to the default `nice` value of 0.

File: nixos/modules/services/backup/restic.nix [hunk 2]
```diff
@@ -256,6 +256,24 @@ in
...
+        };
+
+        ioSchedulingClass = mkOption {
+          description = lib.mdDoc "IO scheduling class for restic (see ionice(1) for a quick description). Applies to local instances, and remote ones connecting by ssh if set to idle. Only used for `restic backup`, `restic forget` and `restic check` commands.";
+          type = types.enum [ "idle" "best-effort" "realtime" ];
```
Comment (oxzi) [inline nixos/modules/services/backup/restic.nix:268]: There is also `none`.

File: nixos/modules/services/backup/restic.nix [hunk 3]
```diff
@@ -256,6 +256,24 @@ in
...
+          default = "best-effort";
```
Comment (oxzi) [inline nixos/modules/services/backup/restic.nix:269]: Same as for my comment regarding the `niceness` default. Maybe default to `none`?

File: nixos/modules/services/backup/restic.nix [hunk 4]
```diff
@@ -256,6 +256,24 @@ in
...
+        };
+
+        ioPriority = mkOption {
+          description = lib.mdDoc "IO scheduling priority for restic (see ionice(1) for a quick description). If null, the value is derived from (niceness + 20) / 5. Only used for `restic backup`, `restic forget` and `restic check` commands.";
```
Comment (oxzi) [inline nixos/modules/services/backup/restic.nix:273]: Might be nitpicking, but could you please align the description's width similar to those of the other options?

File: nixos/modules/services/backup/restic.nix [hunk 5]
```diff
@@ -296,9 +314,21 @@ in
...
+                ioniceClass = {
+                  "idle" = 3;
+                  "best-effort" = 2;
+                  "realtime" = 1;
+                }.${backup.ioSchedulingClass};
+              in
+                ''${pkgs.util-linux}/bin/ionice -t ''
+                + '' -c ${toString ioniceClass} ''
+                + '' ${optionalString (backup.ioPriority != null) "-n ${toString backup.ioPriority}"} ''
+                + '' ${optionalString (backup.niceness >= 1) "${pkgs.coreutils}/bin/nice -n ${toString backup.niceness}"} '';
```
Comment (oxzi) [inline nixos/modules/services/backup/restic.nix:328]: Regarding `backup.niceness >= 1`: Why only allow a positive niceness, when even setting the right restrictions within the option's type. No, I would also expect a negative niceness to work, i.e., when the backup's CPU priority is low.

File: nixos/modules/services/backup/restic.nix [hunk 6]
```diff
@@ -256,6 +256,24 @@ in
...
+
+        ioSchedulingClass = mkOption {
+          description = lib.mdDoc "IO scheduling class for restic (see ionice(1) for a quick description). Applies to local instances, and remote ones connecting by ssh if set to idle. Only used for `restic backup`, `restic forget` and `restic check` commands.";
+          type = types.enum [ "idle" "best-effort" "realtime" ];
+          default = "best-effort";
+        };
+
+        ioPriority = mkOption {
+          description = lib.mdDoc "IO scheduling priority for restic (see ionice(1) for a quick description). If null, the value is derived from (niceness + 20) / 5. Only used for `restic backup`, `restic forget` and `restic check` commands.";
+          type = types.nullOr (types.ints.between 0 7);
```
Comment (oxzi) [inline nixos/modules/services/backup/restic.nix:274]: [ionice(1)](https://linux.die.net/man/1/ionice) states that the "idle" schedule class does not take a priority argument, and only "best effort" and "real time" expect an int within 0-7. Thus, either reflect this restriction with a check or at least add it to the description.

**Review by oxzi — CHANGES_REQUESTED** [review]
Thanks a lot for your PR; I really like the general gist of it.
Could you please address the comments I made? Thanks.

**Review by nagy — COMMENTED** [review]
Could you align your commit messages ( and PR title ) as described in the contribution guidelines?

https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions

With regard to the changes, could you see if you can replicate the same effect with systemd's `Nice` , `IOSchedulingClass` and `IOSchedulingPriority` service config settings ? This could save of from having to mainting this little logic ourselves.

https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html#Scheduling

https://search.nixos.org/options?channel=unstable&show=systemd.services.%3Cname%3E.serviceConfig&from=0&size=50&sort=relevance&type=packages&query=serviceconfig

File: nixos/modules/services/backup/restic.nix (hunk 1 above)
Comment (ibizaman) [reply nixos/modules/services/backup/restic.nix:263]: From https://man7.org/linux/man-pages/man2/setpriority.2.html:

> The prio argument is a value in the range -20 to 19 (but see NOTES below), with -20 being the highest priority and 19 being the lowest priority.  Attempts to set a priority outside this range are silently clamped to the range.  The default priority is 0; lower values give a process a higher scheduling priority.

I agree we should have 0 as the default, if anything just to be backwards compatible. But isn't it the opposite - a higher value means _less_ CPU priority?

**Comment by ibizaman** [comment]
I'm still trying to get the parameters right. I'm wondering if I'm not misunderstanding the values here.

So I've been using the systemd options, it looks like it's going to be better indeed. My PR could transform into just adding something like an `extraServiceConfig` option. Would that be a good idea?

Anyway, the result is having the following set in the systemd config:
```
IOSchedulingClass=best-effort
IOSchedulingPriority=7
Nice=15
```

And this is the output of `ps ax -o pid,ni,pri,pri_baz,class,cmd | grep "restic\|nextcloud"`:
```
    PID  NI PRI BAZ CLS CMD
  46757   0  19 120 TS  php-fpm: pool nextcloud
 154658  15   4 135 TS  /nix/store/fs02cbl7d2q651lpk4i50jp0abv6hi1v-restic-0.16.2/bin/restic backup ....
```
Nextcloud is nothing special here, I just knew this process was running.

So, niceness makes sense, it's 15. But what about PRI? It thought it should be higher than 4?

I'll investigate more on the command line but wanted to let you know where I'm struggling.

**Comment by nagy** [comment]
Maybe it's just easiest to replicate what the nix-daemon service does, see here:

https://github.com/NixOS/nixpkgs/blob/nixos-23.11/nixos/modules/services/system/nix-daemon.nix#L72-L136
and
https://github.com/NixOS/nixpkgs/blob/nixos-23.11/nixos/modules/services/system/nix-daemon.nix#L197-L199

Also be aware, that they dropped the Nice option over there for some reason. 
https://github.com/NixOS/nixpkgs/blob/nixos-23.11/nixos/doc/manual/release-notes/rl-2111.section.md?plain=1#L416

**Comment by ibizaman** [comment]
Thanks for the links. The [commit](https://github.com/NixOS/nixpkgs/commit/ee8e993fd430ed80b1ea58b1b9b396270ade09a1) introducing the change had this message:

> The nix.daemonNiceLevel options allows for setting the nice level of the
Nix daemon process. On a modern Linux kernel with group scheduling the
nice level only affects threads relative to other threads in the same
task group (see sched(7)). Therefore this option has not the effect one
might expect.
> The options daemonCPUSchedPolicy and daemonIOSchedClass are introduced
and the daemonIONiceLevel option renamed to daemonIOSchedPrority for
consistency. These options allow for more effective control over CPU
and I/O scheduling.
> Instead of setting daemonNiceLevel to a high value to increase the
responsiveness of an interactive system during builds -- which would not
have the desired effect, as described above -- one could set both
daemonCPUSchedPolicy and daemonIOSchedClass to idle.

I'm not sure I understand TBH. Specifically:

> the nice level only affects threads relative to other threads in the same
task group (see sched(7)).

I'm not sure if this would produce unexpected results for restic. Also:

> Therefore this option has not the effect one might expect.

I don't know what to expect :sweat_smile:. From this [SO answer](https://stackoverflow.com/a/52501811/1013628), I presume what wasn't clear before the commit was that with nice, a lower number means a higher priority, while with the "sched" priority, the higher the number the higher the priority. It is more intuitive indeed.

@illdefined Hello :wave: I'm pinging you on this PR because of your work on [this commit](https://github.com/NixOS/nixpkgs/commit/ee8e993fd430ed80b1ea58b1b9b396270ade09a1). Could you elaborate on why you switched from `nice` to `sched` for CPU scheduling? I think I understand you wanted a more intuitive relationship with the priority numbers where higher number means higher priority. Is there more to it though? Also, I don't understand how nice values affecting threads relative to other threads in the same task group can lead to weird behaviors. Thanks :)

Btw, I wasn't printing `ionice` values at all. Here is a test that shows `nice` and `ionice` values correctly: (btw if you're wondering what the hell I'm doing with the pipes in there, I actually can't use `$!` to get the pid because in the case of `nice` followed by `ionice`, I can't get the pid of the nested command, I get the pid of `ionice`)
```
$ sleep 2 & pid=$(ps ax -o pid,cmd | grep sleep | grep -v grep | cut -f 2 -d ' '); ionice -p $pid; ps ax -o pid,ni,cmd -q $pid
[1] 716508
none: prio 0
    PID  NI CMD
 716508   0 sleep 2

nice -n 19 sleep 2 & pid=$(ps ax -o pid,cmd | grep sleep | grep -v grep | cut -f 2 -d ' '); ionice -p $pid; ps ax -o pid,ni,cmd -q $pid
[1] 716498
none: prio 0
    PID  NI CMD
 716498  19 sleep 2

ionice -c 2 -n 7 sleep 2 & pid=$(ps ax -o pid,cmd | grep sleep | grep -v grep | cut -f 2 -d ' '); ionice -p $pid; ps ax -o pid,ni,cmd -q $pid
best-effort: prio 7
    PID  NI CMD
 716486   0 sleep 2

nice -n 19 ionice -c 2 -n 7 sleep 2 & pid=$(ps ax -o pid,cmd | grep sleep | grep -v grep | cut -f 2 -d ' '); ionice -p $pid; ps ax -o pid,ni,cmd -q $pid
[2] 716476
best-effort: prio 7
    PID  NI CMD
 716476  19 sleep 2
```

In conclusion, using the systemd options work. I'm just not sure if we should pick `Nice=` or `CPUSchedulingPolicy=`.

**Comment by illdefined** [comment]
> @illdefined Hello 👋 I'm pinging you on this PR because of your work on [this commit](https://github.com/NixOS/nixpkgs/commit/ee8e993fd430ed80b1ea58b1b9b396270ade09a1). Could you elaborate on why you switched from `nice` to `sched` for CPU scheduling? I think I understand you wanted a more intuitive relationship with the priority numbers where higher number means higher priority. Is there more to it though? Also, I don't understand how nice values affecting threads relative to other threads in the same task group can lead to weird behaviors. Thanks :)
> 
> Btw, I wasn't printing `ionice` values at all. Here is a test that shows `nice` and `ionice` values correctly: (btw if you're wondering what the hell I'm doing with the pipes in there, I actually can't use `$!` to get the pid because in the case of `nice` followed by `ionice`, I can't get the pid of the nested command, I get the pid of `ionice`)
> 
> ```
> $ sleep 2 & pid=$(ps ax -o pid,cmd | grep sleep | grep -v grep | cut -f 2 -d ' '); ionice -p $pid; ps ax -o pid,ni,cmd -q $pid
> [1] 716508
> none: prio 0
>     PID  NI CMD
>  716508   0 sleep 2
> 
> nice -n 19 sleep 2 & pid=$(ps ax -o pid,cmd | grep sleep | grep -v grep | cut -f 2 -d ' '); ionice -p $pid; ps ax -o pid,ni,cmd -q $pid
> [1] 716498
> none: prio 0
>     PID  NI CMD
>  716498  19 sleep 2
> 
> ionice -c 2 -n 7 sleep 2 & pid=$(ps ax -o pid,cmd | grep sleep | grep -v grep | cut -f 2 -d ' '); ionice -p $pid; ps ax -o pid,ni,cmd -q $pid
> best-effort: prio 7
>     PID  NI CMD
>  716486   0 sleep 2
> 
> nice -n 19 ionice -c 2 -n 7 sleep 2 & pid=$(ps ax -o pid,cmd | grep sleep | grep -v grep | cut -f 2 -d ' '); ionice -p $pid; ps ax -o pid,ni,cmd -q $pid
> [2] 716476
> best-effort: prio 7
>     PID  NI CMD
>  716476  19 sleep 2
> ```
> 
> In conclusion, using the systemd options works. I'm just not sure if we should pick `Nice=` or `CPUSchedulingPolicy=`.

I think you misunderstand. The nice value gets applied but it does not have the effect you might suspect.

Before group scheduling the nice value would affect the scheduling priority of a task globally. When competing for CPU time, a task with a lower nice value would receive more of it than any other task on the same system with a greater nice value.

But that is no longer the case on most modern Linux systems. With group scheduling, the nice value only affects the scheduling of tasks relative to others in the same task group.

Imagine that you have two single‐threaded processes, one with a nice value of 19 and another one with -20. When both are requiring CPU time, one would expect the one with a lower nice value to receive a lot more CPU time than the one with the greater nice value. But that is only the case if they happen to be in the same task group. If they are in different task groups, they will receive the same amount of CPU time, irrespective of their nice values.

If you are unfamiliar with group scheduling, I suggest you read the section _The nice value and group scheduling_ in the `sched(7)` man page to ensure you understand if adjusting the nice value really has the effect you expect it to have.

File: nixos/modules/services/backup/restic.nix (hunk 1 above)
Comment (oxzi) [reply nixos/modules/services/backup/restic.nix:263]: >But isn't it the opposite - a higher value means _less_ CPU priority?

Yes, of course you are right. I was confused.

Comment (ibizaman) [reply nixos/modules/services/backup/restic.nix:263]: np, it's pretty new to me so I wanted to make sure.
